exporter/exporterhelper/internal/bounded_memory_queue.go

```diff
@@ -17,85 +17,144 @@ import (
 // where the queue is bounded and if it fills up due to slow consumers, the new items written by
 // the producer are dropped.
 type boundedMemoryQueue struct {
-	stopWG       sync.WaitGroup
-	size         *atomic.Uint32
-	stopped      *atomic.Bool
+	size         int
+	eventChan    chan any
 	items        chan Request
-	capacity     uint32
+	capacity     int
 	numConsumers int
+	stopped      *atomic.Bool
+	overflow     *atomic.Bool
 }
 
 // NewBoundedMemoryQueue constructs the new queue of specified capacity, and with an optional
 // callback for dropped items (e.g. useful to emit metrics).
 func NewBoundedMemoryQueue(capacity int, numConsumers int) ProducerConsumerQueue {
 	return &boundedMemoryQueue{
 		items:        make(chan Request, capacity),
-		stopped:      &atomic.Bool{},
-		size:         &atomic.Uint32{},
-		capacity:     uint32(capacity),
+		eventChan:    make(chan any),
+		capacity:     capacity,
 		numConsumers: numConsumers,
+		stopped:      &atomic.Bool{},
+		overflow: func() *atomic.Bool {
+			o := &atomic.Bool{}
+			if capacity == 0 {
+				o.Store(true)
+			}
+			return o
+		}(),
 	}
 }
 
-// StartConsumers starts a given number of goroutines consuming items from the queue
+// Start starts a given number of goroutines consuming items from the queue
 // and passing them into the consumer callback.
 func (q *boundedMemoryQueue) Start(_ context.Context, _ component.Host, set QueueSettings) error {
 	var startWG sync.WaitGroup
```

Comment:
Would it be worth using something like [errgroup.Group](https://pkg.go.dev/golang.org/x/sync/errgroup) instead?

Reply:
I didn't touch that thing :)

Reply:
OK, there's no chance any of the coroutines would return an error, so no, it won't apply here. It's just to force to wait until all coroutines are started.